### PR TITLE
PF-1173: CLI command for GCS object reference type

### PR DESCRIPTION
**Pull request by yuhuyoyo**
Add CLI command for reference type GCS object.

add-ref: It could be a reference to a gcs folder foo/ or a specific file in the folder  foo/bar.txt or every objects in the folder foo/*. If the user have folder level access, they will be able to add a reference to all objects in that folder and a reference to that folder. If the user have only partial access to the folder, they will only be able to add reference to the object that they have access to. 

update: a user can only update the reference if they have access to the referencing target. If they want to update the reference to point to another GCS object, they must also have access to that gcs object. 

describe: if the user has no access to the gcs object, they won't be able to see the detailed gcs blobinfo such as size.

resolve: will resolve to gs://bucketName/path/to/blob or bucketName/path/to/blob depending on whether the gs:// prefix is needed.

### CODE REVIEW

File: src/test/java/unit/GcsBucketFileReferenced.java
Comment (marikomedlock) [inline src/test/java/unit/GcsBucketFileReferenced.java:108]: nit: ...the file is in the list

File: src/test/java/unit/GcsBucketFileReferenced.java
Comment (marikomedlock) [inline src/test/java/unit/GcsBucketFileReferenced.java:174]: nit: suggest making the assertion message different from the previous one so it's easier to see where it failed in test logs (in my experience, the line numbers don't always match up, and certainly won't if there were any changes).

File: src/test/java/unit/GcsBucketFileReferenced.java
Comment (marikomedlock) [inline src/test/java/unit/GcsBucketFileReferenced.java:203]: nit: file

File: src/test/java/unit/GcsBucketFileReferenced.java
Comment (marikomedlock) [inline src/test/java/unit/GcsBucketFileReferenced.java:42]: I think we should test both cases:
- user has access to the bucket, and by extension a file in it
- user does not have full bucket access, only to a specific file in it

No need to duplicate all the tests for both cases, perhaps just for `add-ref` (which does an access check) and `describe` (which will try to lookup information directly from the cloud). If it's easier to put these in a separate class, similar to `GcsBucketNumObjects` unit test class.

File: src/test/java/unit/GcsBucketFileReferenced.java
Comment (marikomedlock) [inline src/test/java/unit/GcsBucketFileReferenced.java:181]: nit: referenced file. ditto for other methods in this class

File: src/test/java/harness/utils/ExternalGCSBuckets.java
Comment (marikomedlock) [inline src/test/java/harness/utils/ExternalGCSBuckets.java:161]: Are there any places other than the overloaded method directly above that specify `Optional.empty()` when calling this method? If not, I think it would be cleaner to not require the `Optional` wrapper here and instead have this method call the one above e.g.:
```
return getGsPath(bucketName) + "/" + fileName;
```

File: src/test/java/unit/GcsBucketFileReferenced.java
Comment (marikomedlock) [inline src/test/java/unit/GcsBucketFileReferenced.java:420]: nit: `listFileResourcesWithName`

File: src/main/java/bio/terra/cli/businessobject/Resource.java
Comment (jmczerk) [reply src/main/java/bio/terra/cli/businessobject/Resource.java:58]: I don't see a lot of context in the Jira ticket (maybe @gjuggler could add color), and I'm not sure if the use of "file" to describe a GCS object is already pervasive across Terra, but...

Hopefully not too pedantic, but I'd lean towards `GCS_OBJECT`/`GcsObject`, since GCS is object storage.  This might be a useful distinction if we ever support actual file storage like [filestore](https://cloud.google.com/filestore), where filesystem semantics are meaningful.

If others think that this is worth  doing, I'd also recommend `filePath` -> `objectName` elsewhere (IME the distinction between object name prefixes and real file system paths is meaningful and can cause confusion for end users).

Comment (jmczerk) [reply src/main/java/bio/terra/cli/businessobject/Resource.java:58]: Actually just noticed that the WSM side is already implemented, so maybe this ship has sailed; feel free to disregard.

File: src/main/java/bio/terra/cli/command/resource/addref/GcsBucketFile.java
Comment (jmczerk) [reply src/main/java/bio/terra/cli/command/resource/addref/GcsBucketFile.java:31]: This is where some of my concern comes from, as folders don't really exist and are really an artificial construct of tools (Cloud Console and `gsutil`):

https://cloud.google.com/storage/docs/folders

I'm not sure we would want to take on managing the facade of folders in Terra.  Using "object name" instead of "file path" semantics sidesteps this issue.

Comment (yuhuyoyo) [reply src/main/java/bio/terra/cli/command/resource/addref/GcsBucketFile.java]: yes, I'm planning to reuse the file referenced resource endpoint in WSM for folder as well. Nichole is interested in having the --file-path to support folder and files. While folder is not really a thing, WSM will just create a reference as long as the access check pass on the bucket level. In long term, we should have a <string, string> map for random things like this. I discussed with Dan and we are happy to settle with this for now to support the feature request.

File: src/main/java/bio/terra/cli/businessobject/Resource.java
Comment (marikomedlock) [reply src/main/java/bio/terra/cli/businessobject/Resource.java]: Object instead of file sounds good to me.

The enum here is CLI-specific, so no need to tie it to WSM naming. One of the main reasons for "redefining" enums here that map to existing WSM client library enums was to let the CLI control display, and I think this fits that case.

File: src/main/java/bio/terra/cli/command/resource/addref/GcsBucketFile.java
Comment (marikomedlock) [reply src/main/java/bio/terra/cli/command/resource/addref/GcsBucketFile.java]: How can WSM tell if a user is referencing a file or folder? Or it doesn't matter because for both we will just check existence/access to the full bucket?

Comment (yuhuyoyo) [reply src/main/java/bio/terra/cli/command/resource/addref/GcsBucketFile.java]: there is a BlobInfo#isDirectory. https://googleapis.dev/java/google-cloud-storage/latest/com/google/cloud/storage/BlobInfo.html#isDirectory--

But you are right, right now we are just checking the access to the full bucket. I'm putting together a PR to change. this PR will wait till that one is merged.

Comment (marikomedlock) [reply src/main/java/bio/terra/cli/command/resource/addref/GcsBucketFile.java]: Ah I didn't know about that `isDirectory` method. Will that require `objects.list` permission on the bucket?

Comment (yuhuyoyo) [reply src/main/java/bio/terra/cli/command/resource/addref/GcsBucketFile.java]: yeah it does require objects.list premission. .. I don't see how to edit folder permission in google cloud console though so I'm not sure how to verify this.

**Review by marikomedlock — COMMENTED** [review]
Following up on a previous conversation in this PR: This CLI command is not intended to support GCS "folders", just blobs right?

File: src/main/java/bio/terra/cli/command/resource/addref/GcsObject.java
Comment (yuhuyoyo) [reply src/main/java/bio/terra/cli/command/resource/addref/GcsObject.java:26]: added example to the description. There's no validation except that if an invalid path is provided, we won't be able to find the. blob and will throw a resource does not exist or you don't have access to it exception in WSM.

File: src/main/java/bio/terra/cli/service/WorkspaceManagerService.java
Comment (yuhuyoyo) [reply src/main/java/bio/terra/cli/service/WorkspaceManagerService.java:904]: the WSM API requires both bucket name and object name to be specified.

File: src/main/java/bio/terra/cli/businessobject/Resource.java
Comment (yuhuyoyo) [reply src/main/java/bio/terra/cli/businessobject/Resource.java:151]: hmm I'm not sure why we need to sync in this case if this method is for updating the common resource properties (name & description).

I added a test for the scenaio you suggested and the test passed. GcsObjectReferenced.updateIndividualProperties.

File: src/main/java/bio/terra/cli/service/WorkspaceManagerService.java
Comment (yuhuyoyo) [reply src/main/java/bio/terra/cli/service/WorkspaceManagerService.java:904]: added support on the CLI side to just update either one.

File: terra
Comment (marikomedlock) [inline terra:2]: Did you mean to commit this file and the terra-cli.tar file? I think these are both generated files.

File: src/test/java/unit/FineGrainedAccessGcsObjectReference.java
Comment (marikomedlock) [inline src/test/java/unit/FineGrainedAccessGcsObjectReference.java:461]: There's no guarantee this method will run in each test method. e.g. if an assertion fails between the `add-user` and `cleanUp` calls. Since you are using the same `shareeUser` in all test methods, you can just grant them the role you want in `setupOnce`.

If you need to test different roles for the `shareeUser`, then I would make a new workspace for each test method that needs this, so that you can be sure the tests won't conflict with each other.

And regarding the resource deletion, it's ok if that doesn't happen for each test because the resources will get cleaned up when the workspace is deleted in `cleanupOnce`. You do need to make sure that each test method uses distinct resource name(s) to avoid collision though.

**Review by marikomedlock — COMMENTED** [review]
Could you please add a test for a folder? (e.g. write a blob "folder1/file1.txt", do not write a blob to "folder1", and then check that we can add that as a reference with both uniform and fine-grained access?). I don't think you are testing that currently, but please correct me if I've missed it.

Also, please add a PR description.

File: src/main/java/bio/terra/cli/command/resource/update/GcsObject.java
Comment (yuhuyoyo) [reply src/main/java/bio/terra/cli/command/resource/update/GcsObject.java:59]: this is to allow user to specify only one properties on the CLI side to update the reference target even though the WSM api requires both.

File: src/main/java/bio/terra/cli/serialization/userfacing/resource/UFGcsObject.java
Comment (yuhuyoyo) [reply src/main/java/bio/terra/cli/serialization/userfacing/resource/UFGcsObject.java:82]: sounds good. I didn't notice that method.

File: src/test/java/unit/FineGrainedAccessGcsObjectReference.java
Comment (yuhuyoyo) [reply src/test/java/unit/FineGrainedAccessGcsObjectReference.java]: done.

**Comment by yuhuyoyo** [comment]
Added test for the case when the user have partial access to a folder/ and test when the user ahve access to the folder due to it being a uniform access bucket.

File: src/main/java/bio/terra/cli/serialization/userfacing/resource/UFGcsObject.java
Comment (yuhuyoyo) [reply src/main/java/bio/terra/cli/serialization/userfacing/resource/UFGcsObject.java:82]: hmm if i changed to getUpdateTime, https://github.com/DataBiosphere/terra-cli/pull/172/files#diff-4565bf9006db6daf2cce9ad6ac8ee4c708353d0534709e10fe81b4e7b06e0e25R395 stop passing.

File: src/main/java/bio/terra/cli/businessobject/Resource.java
Comment (yuhuyoyo) [reply src/main/java/bio/terra/cli/businessobject/Resource.java]: outdated

File: src/main/java/bio/terra/cli/command/resource/update/GcsObject.java
Comment (marikomedlock) [reply src/main/java/bio/terra/cli/command/resource/update/GcsObject.java:59]: ah I see. then I think the right place to include that logic (i.e. WSM API requires both even if they are not being changed) is in `WorkspaceManagerService` when we translate the `UpdateReferencedGcsObjectParams` into the WSM API request object.

File: src/test/java/unit/FineGrainedAccessGcsObjectReference.java
Comment (marikomedlock) [inline src/test/java/unit/FineGrainedAccessGcsObjectReference.java:56]: ok so you have the case where the user has permission only on the blob, and the case where the user has permission on the whole bucket. both of those are good test cases. but what I was trying to say in the comment was to check the case where the user has permission on the folder, but not the whole bucket. e.g. write a single new blob "foo2/secret.txt", grant the `shareeUser` permission on "foo2/" only, then check that the `shareeUser` can add a new reference to "foo2/".

File: src/main/java/bio/terra/cli/serialization/userfacing/resource/UFGcsObject.java
Comment (marikomedlock) [reply src/main/java/bio/terra/cli/serialization/userfacing/resource/UFGcsObject.java:82]: hmm interesting. how far apart are the update times?

File: src/main/java/bio/terra/cli/command/resource/update/GcsObject.java
Comment (yuhuyoyo) [reply src/main/java/bio/terra/cli/command/resource/update/GcsObject.java:59]: hmm it's harder to get the original bucket name or object name in the `WorkspaceManagerService` without calling the resourceApi to get the original bucket object though. do you suggest passing in the original resource  `resource.updateReferenced(gcsObjectParams, resource);`

Comment (marikomedlock) [reply src/main/java/bio/terra/cli/command/resource/update/GcsObject.java:59]: ah good point. How about we add two new fields to the update params class that can hold these properties always: `originalBucketName` and `originalObjectName`? I think that's a little clearer, because otherwise this is inconsistent with all other update classes where `null` means "no change". A comment explaining that this is a limitation of the wsm api would also be helpful.

side note: Why does WSM require this -- is it possible to change that in the future?

File: src/main/java/bio/terra/cli/serialization/userfacing/resource/UFGcsObject.java
Comment (yuhuyoyo) [reply src/main/java/bio/terra/cli/serialization/userfacing/resource/UFGcsObject.java:82]: I added some logging. It's 6 seconds apart.

sharedblob update time = 2021-12-20T14:34:41.500+0000
actual update time = 2021-12-20T14:34:47.633+0000

File: src/main/java/bio/terra/cli/command/resource/update/GcsObject.java
Comment (yuhuyoyo) [reply src/main/java/bio/terra/cli/command/resource/update/GcsObject.java:59]: Yeah. I didn't realize this Api causes inconvenience from the client side. We could change it on the WSM side in the future. Currently the `UpdateGcsBucketObjectReferenceRequestBody` reuses the `GcpGcsObjectAttributes` which requires both bucket name and object name to be present. `GcpGcsObjectAttributes` was originally used for creating a reference to the bucket object which makes sense to have both field to be present.

File: src/test/java/unit/GcsObjectReferenced.java
Comment (yuhuyoyo) [reply src/test/java/unit/GcsObjectReferenced.java:233]: it wouldn't fail actually because the user have access to everything in the blobs/ folder. 

Currently in WSM the check access code use something like `storage.get(bucketname, blobId)` to get the blob. If get succeeded and no exception is thrown, we know the user have at least READER access. However, if the blobId is "blobs/**", "blobs/*.txt", "blobs/*.png", there's actually no such object in the storage, so storage.get() will succeed but will return null. Currently checkAccess will still pass for these cases to be more flexible on what to pass in.

File: src/test/java/unit/FineGrainedAccessGcsObjectReference.java
Comment (yuhuyoyo) [reply src/test/java/unit/FineGrainedAccessGcsObjectReference.java:56]: i actually can't grant access to the foo2/ bucket without creating it. 

When I was adding the integration test in WSM, I used the console UI to create the folder and added object in the folder and then grant distinct access to the folder and the object.

File: src/test/java/unit/GcsObjectReferenced.java
Comment (marikomedlock) [reply src/test/java/unit/GcsObjectReferenced.java:233]: Would "blobs1234/" fail since that folder doesn't exist? Or can you give an example that should fail?

File: src/test/java/unit/FineGrainedAccessGcsObjectReference.java
Comment (marikomedlock) [reply src/test/java/unit/FineGrainedAccessGcsObjectReference.java:56]: sorry, not completely following. If there is just a folder, and no object (e.g. "gs://mybucket/foo2/secret.txt" is the only object, folder is "foo2"), is it possible to grant access to just that folder, either through the console UI or Java client library? or do you always have to grant access to an object?

Comment (yuhuyoyo) [reply src/test/java/unit/FineGrainedAccessGcsObjectReference.java:56]: folder is an object. so if i create a foo2/ folder, i can grant access to foo2/. The console uI doesn't let you grant access to a folder but you can do that through gsutil in command line. 

the issue here is that if I programmatically created a foo2/bar.txt object, cloud storage doesn't automatically create a foo2/ object as well. so if I try to grant access to foo2/, this object does not exist.

Comment (marikomedlock) [reply src/test/java/unit/FineGrainedAccessGcsObjectReference.java:56]: can you clarify what you mean by "create a foo2/" folder? do you mean to write a non-zero size blob to "gs://mybucket/foo2/"?

Comment (yuhuyoyo) [reply src/test/java/unit/FineGrainedAccessGcsObjectReference.java:56]: It can be zero-bite.
gcloud documentation: https://cloud.google.com/storage/docs/folders#overview

"If you create an empty folder using the Cloud Console, Cloud Storage creates a zero-byte object as a placeholder. For example, if you create a folder called folder in a bucket called my-bucket, a zero- byte object called gs://my-bucket/folder/ is created. You can find this object when you run gsutil ls."

File: src/test/java/unit/GcsObjectReferenced.java
Comment (yuhuyoyo) [reply src/test/java/unit/GcsObjectReferenced.java:233]: i believe it won't fail either because this log-in user has bucket level access so they can perform get action to anything in this bucket. 

I set the check very lenient because Nichole is interested in having a reference to foo/**, foo/**.txt, etc which are not actual objects but we can tell the access level roughly by checking if the user can perform get action inside folder foo/ regardless of whether there's anything in that folder. We could make the check stricter on the WSM side but I think that's outside the scope of this PR.

**Review by marikomedlock — COMMENTED** [review]
ok so to summarize for GCS referenced resources:
- WSM will check that the user has access either to the object (if it exists) or the whole bucket (if it does not)
- WSM will not check that the object exists

I think the main difference now is that we are no longer guaranteeing existence of GCS objects. that's different than the behavior of other referenced resource types where we check both existence and access. is that an accurate summary?

if so, then we should document this (i.e. GCS behavior and how it's different from other resource types). perhaps a new sub-header in the README under `Resources`, similar to the [section](https://github.com/DataBiosphere/terra-cli/blob/main/README.md#gcs-bucket-lifecycle-rules) on GCS lifecycle rules would be good.

**Comment by yuhuyoyo** [comment]
Almost.

WSM will check that the user has at least READ access to the object (if it exists) or the path contains a folder object that the user has READ access to (e.g. foo/bar/*) where the user may have no full access to the bucket but access to foo/bar/. 

added description in README

File: src/test/java/unit/GcsObjectReferenced.java
Comment (marikomedlock) [inline src/test/java/unit/GcsObjectReferenced.java:234]: Is this leftover from debugging?

File: README.md
Comment (marikomedlock) [inline README.md:482]: nit: maybe remove the "but because of"?

File: README.md
Comment (marikomedlock) [inline README.md:464]: nit: for this and below, please make it a new section (i.e. start with "#####" and add it to the table of contents at the top).

**Review by marikomedlock — APPROVED** [review]
LGTM. Thanks for all the clarification! 3 nits included in comments, please take a look before merging.

**Review by jmczerk — APPROVED** [review]
One small documentation nit, otherwise LGTM

File: README.md
Comment (yuhuyoyo) [reply README.md:485]: it is actually an intended typo. Because we do not check whether the object exist, despite there's a foo/ object and not a fooo/ object, a reference to fooo/ can be created if the user have access to the specified bucket. So when using this command, user needs to pay attention to the object name.

Comment (jmczerk) [reply README.md:485]: OK, it might be worth saying something like "(where object `fooo` does not exist)" to clarify, but I'll leave that up to you.

**Pull request merged**
